egs/librispeech/ASR/transducer/rnn.py

```diff
@@ -0,0 +1,652 @@
+          state:
+            If not ``None``, it contains the hidden state (h, c) for each
+            element in the batch. Both are of shape (batch_size, hidden_size)
+            if proj_size is 0. If proj_size is not zero, the shape of `h` is
+            (batch_size, proj_size).
+            If ``None``, it uses zeros for `h` and `c`.
+        Returns:
+          Return two tensors:
+            - `next_h`: It is of shape (batch_size, hidden_size) if proj_size
+              is 0, else (batch_size, proj_size), containing the next hidden
+              state for each element in the batch.
+            - `next_c`: It is of shape (batch_size, hidden_size) containing the
+              next cell state for each element in the batch.
+        """
+        if state is None:
+            zeros = torch.zeros(
+                input.size(0),
+                self.hidden_size,
+                dtype=input.dtype,
+                device=input.device,
+            )
+            state = (zeros, zeros)
+
+        hx, cx = state
+        gates = F.linear(input, self.weight_ih, self.bias_ih) + F.linear(
+            hx, self.weight_hh, self.bias_hh
+        )
+
+        in_gate, forget_gate, cell_gate, out_gate = gates.chunk(chunks=4, dim=1)
+
+        in_gate = self.layernorm_i(in_gate)
+        forget_gate = self.layernorm_f(forget_gate)
+        cell_gate = self.layernorm_cx(cell_gate)
+        out_gate = self.layernorm_o(out_gate)
+
+        in_gate = torch.sigmoid(in_gate)
+        forget_gate = torch.sigmoid(forget_gate)
+        cell_gate = torch.tanh(cell_gate)
+        out_gate = torch.sigmoid(out_gate)
+
+        cy = (forget_gate * cx) + (in_gate * cell_gate)
+        cy = self.layernorm_cy(cy)
+        hy = out_gate * torch.tanh(cy)
+
+        if self.weight_hr is not None:
+            hy = torch.matmul(hy, self.weight_hr.t())
+
+        return hy, cy
+
+    def extra_repr(self) -> str:
+        s = "{input_size}, {hidden_size}"
+        if "bias" in self.__dict__ and self.bias is not True:
+            s += ", bias={bias}"
+        return s.format(**self.__dict__)
+
+    def reset_parameters(self) -> None:
+        stdv = 1.0 / math.sqrt(self.hidden_size)
+        for name, weight in self.named_parameters():
+            if "layernorm" not in name:
+                nn.init.uniform_(weight, -stdv, stdv)
+
+
+class LayerNormLSTMLayer(nn.Module):
+    """
+    Examples::
+
+        >>> layer = LayerNormLSTMLayer(10, 20)
+        >>> input = torch.rand(2, 5, 10)
+        >>> h0 = torch.rand(2, 20)
+        >>> c0 = torch.rand(2, 20)
+        >>> output, (hn, cn) = layer(input, (h0, c0))
+        >>> output.shape
+        torch.Size([2, 5, 20])
+        >>> hn.shape
+        torch.Size([2, 20])
+        >>> cn.shape
+        torch.Size([2, 20])
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool = True,
+        ln: nn.Module = nn.LayerNorm,
```

Comment:
nn.LayerNorm is a class object, not a class instance, so I think it should be `ln: class`.
I'm not sure how we can test that these type annotations are correct.

Reply:
Ah, thanks.

There are some type checking tools (not mypy).

Will add it.

Reply:
It's `Type[nn.Module]` [(`from typing import Type`)](https://docs.python.org/3.8/library/typing.html#typing.Type)

Reply:
It's `Type[nn.Module]` [(`from typing import Type`)](https://docs.python.org/3.8/library/typing.html#typing.Type)

Reply:
thanks